### PR TITLE
Code Editor

### PR DESCRIPTION
**Pull request by yung-coder**
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

# Description

So, here I have tried to create a code editor for HTML CSS, and JS using codemirror 

Fixes # (issue)
#453 
## Type of change

Please delete options that are not relevant.

- [ ✔️] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
So, it has been tested by me but there is a problem with duplicating  editors it is a kind of a bug I have already told about this to @atapas it would be great if anyone can help to resolve this bug

It is working totally fine the only problem is this extra editor 🔽

ink

# Checklist:

- [ ✔️] I have performed a self-review of my own code
- [✔️ ] I have commented my code, particularly in hard-to-understand areas
- [ ✔️] I have made corresponding changes to the documentation
- [✔️ ] My changes generate no new warnings
- [ ✔️] I have added tests that prove my fix is effective or that my feature works
- [ ✔️] New and existing unit tests pass locally with my changes
- [✔️ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/plays/code-editor/styles.css [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/* enter stlyes here */
```
Comment (atapas) [inline src/plays/code-editor/styles.css:1]: Can you please prefix a unique class name to all the styles?

**Review by atapas — CHANGES_REQUESTED** [review]
@yung-coder Started review and initial comments.

**Comment by atapas** [comment]
@yung-coder I have noticed that the build is not going through for this branch. Could you please check what's different?

Here is the build error:



4/4] Building fresh packages...
--
12:34:30.884 | success Saved lockfile.
12:34:30.889 | Done in 84.21s.
12:34:31.442 |  
12:34:31.442 | > react-play@1.0.0 build
12:34:31.442 | > npx create-react-play -p && react-scripts build
12:34:31.442 |  
12:34:33.417 | /usr/bin/env: node : No such file or directory
12:34:33.431 | Error! Command "CI= npm run build && npm run snap" exited with 127

File: package.json [hunk 2]
```diff
@@ -5,6 +5,9 @@
   "dependencies": {
     "@emotion/react": "^11.9.3",
     "@emotion/styled": "^11.9.3",
+    "@fortawesome/fontawesome-svg-core": "^6.1.2",
```
[the remark this answers is not in the document]
Comment (yung-coder) [reply package.json:8]: Okay i will

File: src/plays/code-editor/CodeEditor.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+              <Editor
+                language="javascript"
+                displayName="JS"
+                value={js}
+                setEditorState={setJs}
+              />
+            )}
+          </div> 
+          <div className="bottom-pane">
+            <iframe
```
[the remark this answers is not in the document]
Comment (yung-coder) [reply src/plays/code-editor/CodeEditor.js:91]: It's for displaying the real time output of the code using codemirror

File: package.json [hunk 4]
```diff
@@ -23,6 +28,7 @@
     "lodash": "^4.17.21",
     "node-sass": "^7.0.1",
     "react": "^18.0.0",
+    "react-codemirror2": "^7.2.1",
```
[the remark this answers is not in the document]
Comment (yung-coder) [reply package.json:31]: Yes

File: package.json [hunk 5]
```diff
@@ -13,6 +16,8 @@
     "@types/react-dom": "^18.0.2",
     "add": "^2.0.6",
     "axios": "^0.27.2",
+    "codemirror": "^5.65.7",
+    "create-react-play": "^0.0.10",
```
[the remark this answers is not in the document]
Comment (yung-coder) [reply package.json:20]: So when i was creating this play it asked me to install this package...

File: src/plays/code-editor/styles.css (hunk 1 above)
Comment (yung-coder) [reply src/plays/code-editor/styles.css:1]: Okay

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (atapas) [reply package.json:20]: Shouldn't. @koustov are we not doing npx for it?

[the remark this answers is not in the document]
Comment (koustov) [reply package.json:20]: Yes, we shouldn't use `create-react-play` as a project dependency. It should be global and we are running it through `npx`

File: src/plays/code-editor/styles.css (hunk 1 above)
Comment (yung-coder) [reply src/plays/code-editor/styles.css:1]: Do you want me to change all the class names or a specific one ...?

Comment (atapas) [reply src/plays/code-editor/styles.css:1]: All, what I meant is, add a unique class prefix like,

.code-editor .existing-class

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (atapas) [reply package.json:20]: @yung-coder remove this one.

[the remark this answers is not in the document]
Comment (yung-coder) [reply package.json:20]: Okay

File: src/plays/code-editor/styles.css (hunk 1 above)
Comment (yung-coder) [reply src/plays/code-editor/styles.css:1]: Okay

**Review by koustov — CHANGES_REQUESTED** [review]
the cover image should be in png format

**Comment by koustov** [comment]
@yung-coder the build is passing now as you have removed the package.

**Comment by yung-coder** [comment]
Okay i will change the cover image give a minute 🕜

**Comment by yung-coder** [comment]
@koustov is 30 kb enough..?

**Review by atapas — CHANGES_REQUESTED** [review]
A few feedback:

1. The button text for selecting a language is not so readable with blue background and black font color.
2. Can you create a section for the output? That clearly shows where the output is expected when code is typed?
3. One issue I faced was, when I was typing the code in my usual speed, the cursor was always going one character backwards.

Could you please look into these issues?

File: src/plays/code-editor/Readme.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Code editor 
+
+It is a code editor for html css and javascript it is like a clone of codepen !
```
Comment (atapas) [inline src/plays/code-editor/Readme.md:3]: Please add more technical details as it will help many.

**Comment by yung-coder** [comment]
Hey @atapas i have made the changes but after that i had merge the main branch so do i have to npm new packages...?

**Comment by yung-coder** [comment]
@atapas any further modifications..?

**Comment by atapas** [comment]
> @atapas any further modifications..?

Code wise all good. A few modification can be done to improve the look and feel. Except comments today.

**Comment by atapas** [comment]
@yung-coder 

I have a few questions regarding the code editor
- Will the JavaScript output come in the output section? Like if I do a console log? Or will it appear in the browser's console? If so, better to provide a note about it when someone is in the JavaScript Tab.
- Isn't it good to clear the output area when user switch between the tabs?
- The above question may not be valid if we expect the user to try HTML, CSS, and JS to type in each section and it together outputs it. Please let me know what's the expectation.
- How to see the CSS output?

**Comment by yung-coder** [comment]
@atapas 

No javascript output will not come in the output section but we can manipulate things through it as usual and about clearing tabs on switching is not possible as it together gives the output css output is also visible in the output section as  we try to make style changes....!

Hope i answered all your questions
if you want to know more or to make any modifications please let me know..!

**Comment by atapas** [comment]
@yung-coder is it ready for the final review?

**Comment by yung-coder** [comment]
Yes i have added the modal please check it if any changes to make or some content changes....!

**Comment by atapas** [comment]
Hey the how to play is still shows in black color to me and does not appear like a link. Also can we change the text as “How To Use?”

**Comment by yung-coder** [comment]
@atapas in which color to change it ...

**Comment by yung-coder** [comment]
Hey @atapas  so I have made the necessary changes and I think it's all good now ....

**Review by atapas — APPROVED** [review]

**Pull request merged**